### PR TITLE
SH-104, SH-170 and fixed dependency management

### PR DESCRIPTION
**Pull request by sherzinger**

### CODE REVIEW

File: web-app/js/smartR/_angular/controllers/heatmap.js [hunk 1]
```diff
@@ -8,179 +8,115 @@ function HeatmapController ($scope, smartRUtils, commonWorkflowService) {
...
+                runAnalysisModel.scriptResults = {};
+            }
+
+            // clear old results
+            if (preprocessRunning) {
+                runAnalysisModel.scriptResults = {};
+            }
+
+            // disable tabs when certain criteria are not met
+            fetchModel.tab.disabled = preprocessRunning || runAnalysisRunning;
```
Comment (rnugraha) [inline web-app/js/smartR/_angular/controllers/heatmap.js:98]: I would not recommend this. We got input from users that they still wanna be able to view and configure options for next process while something is running in the background. For example, while fetching the data (which is normally long) I want to be able to prepare options.

File: web-app/js/smartR/_angular/controllers/heatmap.js [hunk 2]
```diff
@@ -8,179 +8,115 @@ function HeatmapController ($scope, smartRUtils, commonWorkflowService) {
...
+
+            // disable buttons when certain criteria are not met
+            fetchModel.btn.disabled = !conceptBoxesHighDimensionalValid;
+            runAnalysisModel.download.disabled = runAnalysisRunning || $.isEmptyObject(scope.runAnalysis.scriptResults);
+
+            // set ranking criteria
+            runAnalysisModel.subsets = scope.fetch.subsets;
+            if (scope.fetch.subsets < 2 &&
+                    ['logfold', 'bval', 'pval', 'adjpval', 'ttest'].indexOf(runAnalysisModel.params.ranking) !== -1) {
+                runAnalysisModel.params.ranking = 'mean';
```
Comment (rnugraha) [inline web-app/js/smartR/_angular/controllers/heatmap.js:110]: if subsets < 2 ---> rangking = 'coef'
if number of samples == 1 --> rangking = 'mean'

File: web-app/js/smartR/_angular/directives/fetchButton.js [hunk 3]
```diff
@@ -8,6 +8,8 @@ window.smartRApp.directive('fetchButton',
         return {
             restrict: 'E',
             scope: {
+                loaded: '=',
```
Comment (rnugraha) [inline web-app/js/smartR/_angular/directives/fetchButton.js]: I'm not sure about this. The dangerous situation would be, would this 'loaded' flag could be trustable, while the content could be empty.

File: web-app/js/smartR/_angular/controllers/heatmap.js (hunk 1 above)
Comment (sherzinger) [inline web-app/js/smartR/_angular/controllers/heatmap.js:98]: The problem is that switching between tabs can cause graphic glitches when drawing is happening in an inactive tab. They will have to wait or deal with bugs.

File: web-app/js/smartR/_angular/directives/fetchButton.js (hunk 3 above)
Comment (sherzinger) [inline web-app/js/smartR/_angular/directives/fetchButton.js]: This is far more safe than binding it to the button state, which is what happens right now.

File: web-app/js/smartR/_angular/controllers/heatmap.js (hunk 1 above)
Comment (rnugraha) [inline web-app/js/smartR/_angular/controllers/heatmap.js:98]: I personally don't mind with that but if we have good reason, this should be mentioned as well in the Changelog so it wont surprise users.

**Pull request merged**